storage/rocksdb/rdb_datadic.h

```diff
@@ -1354,4 +1368,82 @@ struct Rdb_index_info {
   uint64 m_ttl_duration = 0;
 };
 
+/*
+  @brief
+  Merge Operator for the auto_increment value in the system_cf
+
+  @detail
+  This class implements the rocksdb Merge Operator for auto_increment values
+  that are stored to the data dictionary every transaction.
+
+  The actual Merge function is triggered on compaction, memtable flushes, or
+  when get() is called on the same key.
+
+ */
+class Rdb_merge_operator : public rocksdb::AssociativeMergeOperator {
+ public:
+  /*
+    Updates the new value associated with a key to be the maximum of the
+    passed in value and the existing value.
+
+    @param[IN]  key
+    @param[IN]  existing_value  existing value for a key; nullptr if nonexistent
+    key
+    @param[IN]  value
+    @param[OUT] new_value       new value after Merge
+    @param[IN]  logger
+  */
+  bool Merge(const rocksdb::Slice &key, const rocksdb::Slice *existing_value,
+             const rocksdb::Slice &value, std::string *new_value,
+             rocksdb::Logger *logger) const override {
+
+    /* value corruption check and version check */
+    DBUG_ASSERT(new_value != nullptr);
+    DBUG_ASSERT(FindVersion(value) == Rdb_key_def::AUTO_INCREMENT_VERSION);
+
+    std::string merged_value;
+
+    /* if no existing vaue, then persist the passed in value */
+    if (existing_value == nullptr) {
+      merged_value = Serialize(Deserialize(value));
+    } else {
+      merged_value =
+          Serialize(std::max(Deserialize(value), Deserialize(*existing_value)));
+    }
+
+    *new_value = merged_value;
+    return true;
+  }
+
+  virtual const char *Name() const override {
+    return "system_cf_auto_increment_merge";
```

Comment:
I was thinking the class itself: Rdb_merge_operator, is specific to the __system__ column family. Maybe it can be renamed too? Rdb_system_merge_op. Adding "auto_increment" is probably too specific here. The merge operator can also be used to perform other types of updates with in the __system__ column family. Auto increment is just one of them.

Hence, why a key check should be performed within the Merge() routine.